### PR TITLE
Improve splunk logging of unsuccessful login attempts

### PR DESCRIPTION
**Pull request by barrucadu**
**Record user agent string on unsuccessful login:** done by adding a `user_agent_string` field to the `event_logs` table.  I'm not sure if this approach is better or worse than creating `UserAgent` models, I went for this because I figured it would be pretty straightforward to prune the event log if it grows too big.  If there are too many `UserAgent` values, then both `event_logs` and `user_agents` need pruning.

**Record email address when no such account exists:** done by making the `uid` field of `event_logs` optional (though it's required for all pre-existing event types), adding a `user_email_string` field, and adding a new event type.

The new event is type 38 and passes the email address in the `user` field.

---

https://govuk.zendesk.com/agent/tickets/3783119
https://govuk.zendesk.com/agent/tickets/3788622

### CODE REVIEW

File: app/models/event_log.rb [hunk 1]
```diff
@@ -88,7 +88,7 @@ def send_to_splunk(*)
       src_ip: (self.ip_address_string if self.ip_address.present?),
       action: self.event,
       result: self.trailing_message,
-      http_user_agent: self.user_agent&.user_agent_string
+      http_user_agent: self.user_agent&.user_agent_string || self.user_agent_string,
```
Comment (sihugh) [inline app/models/event_log.rb]: Is this the same thing again or did I miss something? `user_agent_string` appears to be delegated to `user_agent`:  https://github.com/alphagov/signon/blob/5b66fcbf425e35035bbafeee4fafdc3243b235d7/app/models/event_log.rb#L65

Comment (barrucadu) [reply app/models/event_log.rb]: Whoops, that would have been a bug.  Good catch

File: app/models/event_log.rb [hunk 2]
```diff
@@ -6,53 +6,55 @@ class EventLog < ActiveRecord::Base
...
-    ACCESS_TOKEN_GENERATED = LogEntry.new(id: 29, description: "Access token generated", require_application: true, require_initiator: true),
-    ACCESS_TOKEN_REVOKED = LogEntry.new(id: 30, description: "Access token revoked", require_application: true, require_initiator: true),
-    PASSWORD_RESET_LOADED_BUT_TOKEN_EXPIRED = LogEntry.new(id: 31, description: "Password reset page loaded but the token has expired"),
-    SUCCESSFUL_PASSWORD_RESET = LogEntry.new(id: 32, description: "Password reset successfully"),
-    ROLE_CHANGED = LogEntry.new(id: 33, description: "Role changed", require_initiator: true),
-    ACCOUNT_UPDATED = LogEntry.new(id: 34, description: "Account updated", require_initiator: true),
-    PERMISSIONS_ADDED = LogEntry.new(id: 35, description: "Permissions added", require_initiator: true),
-    PERMISSIONS_REMOVED = LogEntry.new(id: 36, description: "Permissions removed", require_initiator: true),
-    ACCOUNT_INVITED = LogEntry.new(id: 37, description: "Account was invited", require_initiator: true),
+    ACCOUNT_LOCKED = LogEntry.new(id: 1, description: "Password verification failed too many times, account locked for #{LOCKED_DURATION}", require_uid: true),
```
Comment (erino) [inline app/models/event_log.rb]: I wonder if the default for `require_uid ` should be `true` as nearly all of these `LogEntry`s have `require_uid: true`.

Comment (barrucadu) [reply app/models/event_log.rb]: I thought about that, but it would be inconsistent with the others, which could be confusing

Comment (erino) [reply app/models/event_log.rb]: Fair enough

**Pull request merged**